lib/util/DocumentExtractor.js

```diff
@@ -0,0 +1,158 @@
+      insertInRequest: (documents, request) => {
+        request.input.body.ids = [];
+        let it = 0; // NOSONAR
+        for (; it < documents.length; it += 1) {
+          request.input.body.ids.push(documents[it]._id);
+        }
+        return request;
+      },
+      insertInResult: (documents, request) => {
+        if (request.input.action === 'mGet') {
+          request.setResult({}, { status: request.status });
+          request.result.hits = documents;
+          request.result.total = request.result.hits.length;
+          return request;
+        }
+        const result = [];
+        let it = 0; // NOSONAR
+        for (; it < documents.length; it += 1) {
+          result.push(documents[it]._id);
+        }
+        request.setResult(result, { status: request.status });
+        return request;
+      }
+    }
+  },
+  {
+    targets: ['mCreate', 'mCreateOrReplace', 'mReplace', 'mUpdate'],
+    methods: {
+      extractFromRequest: request => {
+        const documents = [];
+        let it = 0; // NOSONAR
+        for (; it < request.input.body.documents.length; it += 1) {
+          const { _id, body: _source } = request.input.body.documents[it];
+          documents.push({ _id, _source });
+        }
+        return documents;
+      },
+      extractFromResult: request => request.result.hits,
+      insertInRequest: (documents, request) => {
+        request.input.body.documents = [];
+        let it = 0; // NOSONAR
+        for (; it < documents.length; it += 1) {
+          const { _id, _source: body } = documents[it];
+          request.input.body.documents.push({ _id, body});  
+        }
+        return request;
+      },
+      insertInResult: (documents, request) => {
+        request.setResult({}, { status: request.status });
+        request.result.hits = documents;
+        request.result.total = request.result.hits.length;
+        return request;
+      }
+    }
+  },
+  {
+    targets: ['create', 'createOrReplace', 'replace', 'update', '_default'],
+    methods: {
+      extractFromRequest: request => [{ _id: request.input.resource._id, _source: request.input.body }],
+      extractFromResult: request => [request.result],
+      insertInRequest: (documents, request) => {
+        request.input.resource._id = documents[0] && documents[0]._id;
+        request.input.body = documents[0] && documents[0]._source;
+        return request;
+      },
+      insertInResult: (documents, request) => {
+        request.setResult(documents[0], { status: request.status });
+        return request;
+      }
+    }
+  }
+]).reduce((acc, extractor) => {
+  extractor.targets.forEach(target => {
+    acc[target] = extractor.methods;
+  });
+  return acc;
+}, {});
+
+class DocumentExtractor {
+  constructor(request) {
+    this.request = request;
+
+    const extractor = extractors[request.input.action] || extractors._default;
+    this.extractMethod = extractor[request.result ? 'extractFromResult' : 'extractFromRequest'];
```

Comment:
(nitpicking) I think DRYifying the test on `request.result` would make it (marginally) easier to read.